From 063f09a0d7310015e3f0ac22a47b27bb9cfd0a2d Mon Sep 17 00:00:00 2001
From: Rashed Abdel-Tawab <rashed@linux.com>
Date: Tue, 31 Oct 2017 19:58:27 -0700
Subject: [PATCH 12/25] camera: Allow devices to load custom CameraParameter
 code

* Some devices need additional code to load their cameras. Add a hook
  for extra classes and symbols to be included.

Originally implemented in Android.mk at 38b27bf2e5079441358288402b38865ba7c1d9d6
Credit to Ethan Chen for original Android.mk implementation

Change-Id: Iec667af2f0bfdb41f14df342cb128908244d7af0
---
 camera/Android.bp    | 32 ++++++++++++++++++++++++++-
 camera/parameters.go | 52 ++++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 83 insertions(+), 1 deletion(-)
 create mode 100644 camera/parameters.go

diff --git a/camera/Android.bp b/camera/Android.bp
index 24b391856..7cc0336c3 100644
--- a/camera/Android.bp
+++ b/camera/Android.bp
@@ -14,6 +14,26 @@
 
 subdirs = ["ndk"]
 
+bootstrap_go_package {
+    name: "soong-camera-parameters",
+    pkgPath: "android/soong/camera_parameters",
+    deps: [
+        "blueprint",
+        "blueprint-pathtools",
+        "soong",
+        "soong-android",
+        "soong-cc",
+    ],
+    srcs: [
+        "parameters.go",
+    ],
+    pluginFor: ["soong_build"],
+}
+
+camera_parameters_defaults {
+    name: "camera_parameters_defaults",
+}
+
 cc_library_shared {
     name: "libcamera_client",
 
@@ -34,7 +54,6 @@ cc_library_shared {
         // Source for camera interface parcelables, and manually-written interfaces
         "Camera.cpp",
         "CameraMetadata.cpp",
-        "CameraParameters.cpp",
         "CaptureResult.cpp",
         "CameraParameters2.cpp",
         "ICamera.cpp",
@@ -49,6 +68,17 @@ cc_library_shared {
         "VendorTagDescriptor.cpp",
     ],
 
+    product_variables: {
+        lineage: {
+            uses_generic_camera_parameter_library: {
+                srcs: [
+                    "CameraParameters.cpp",
+                ],
+            },
+        },
+    },
+    defaults: ["camera_parameters_defaults"],
+
     shared_libs: [
         "libcutils",
         "libutils",
diff --git a/camera/parameters.go b/camera/parameters.go
new file mode 100644
index 000000000..dd12167e1
--- /dev/null
+++ b/camera/parameters.go
@@ -0,0 +1,52 @@
+// Copyright (C) 2017 The Android Open Source Project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package lineage;
+
+import (
+  "android/soong/android"
+  "android/soong/cc"
+)
+
+func init() {
+  android.RegisterModuleType("camera_parameters_defaults", cameraParametersFactory)
+}
+
+func cameraParametersFactory() android.Module {
+  module := cc.DefaultsFactory()
+  android.AddLoadHook(module, cameraParameters)
+
+  return module
+}
+
+func cameraParameters(ctx android.LoadHookContext) {
+  type props struct {
+    Whole_static_libs []string
+  }
+
+  p := &props{}
+  p.Whole_static_libs = globalDefaults(ctx)
+
+  ctx.AppendProperties(p)
+}
+
+func globalDefaults(ctx android.BaseContext) ([]string) {
+  var staticLibs []string
+
+  device_camera_parameters_lib := ctx.DeviceConfig().SpecificCameraParametersLibrary()
+  if (len(device_camera_parameters_lib) > 0) {
+    staticLibs = append(staticLibs, device_camera_parameters_lib)
+  }
+
+  return staticLibs
+}
-- 
2.17.1

